[server] Integrate landingpage build process with cmake

pathmapper opened the pull request:
## Description

closes https://github.com/qgis/QGIS/issues/46040

This is a first attempt to integrate the landingpage build process with cmake so there's no need anymore to compile and update the compiled js files manually.

This approach is using a custom target which depends on a file created by a custom command. This file is currently just a timestamp and not an `md5` hash like shown in the link of https://github.com/qgis/QGIS/issues/46040#issuecomment-975648298, because it may be sufficient that `yarn install` handles the dependencies.

For the first build the dependencies are installed and the js files are compiled, afterwards only the files are compiled if there's a change to the landingpage sources.

Using this approach locally workes fine so far, but not sure if this is the way to go. 

It's a bit annoying that the status of the build progress is logged several times among other cmake logging output, maybe this could be fixed somehow.    

Submitting this PR as a draft for feedback/discussion and also to see if and how the CI build is going.

pathmapper committed:
Integrate landingpage build process with cmake

m-kuhn commented on resources/CMakeLists.txt:
```diff
@@ -32,6 +32,24 @@ endif()
+    OUTPUT ${CMAKE_CURRENT_SOURCE_DIR}/server/src/landingpage/landingpage.stamp
+    DEPENDS ${LANDINGPAGE_SOURCE_FILES}
+    COMMAND ${CMAKE_COMMAND} -E touch ${CMAKE_CURRENT_SOURCE_DIR}/server/src/landingpage/landingpage.stamp
+    COMMAND cd ${CMAKE_CURRENT_SOURCE_DIR}/server/src/landingpage && yarn install --frozen-lockfile && yarn build
```
Would it be possible to copy first and build after, this way the source tree would be left untouched
Same for the landingpage.stamp, this should go into the build dir and not into the source dir.

pathmapper replied:
Good idea! Yes, that's possible, done in https://github.com/qgis/QGIS/pull/46643/commits/34f7d640c5b5b0f4b19472906b503371685b347b.

pathmapper committed:
Copy first and build in build directory

3nids commented on resources/CMakeLists.txt:
```diff
@@ -32,6 +32,24 @@ endif()
+  file(GLOB_RECURSE LANDINGPAGE_SOURCE_FILES RELATIVE ${CMAKE_CURRENT_SOURCE_DIR} server/src/landingpage/*)
+  set(LANDINGPAGE_OUTPUT_PATH ${CMAKE_CURRENT_BINARY_DIR}/../output/data/resources/server/src/landingpage)
+
+  add_custom_target (yarn-target ALL DEPENDS ${LANDINGPAGE_OUTPUT_PATH}/landingpage.stamp)
```
can we have a more intuitive name for the target?
```suggestion
  add_custom_target (server_landingpage ALL DEPENDS ${LANDINGPAGE_OUTPUT_PATH}/landingpage.stamp)
```
?

pathmapper replied:
:+1: sure

3nids commented on resources/CMakeLists.txt:
```diff
@@ -32,6 +32,24 @@ endif()
+
+  # Landingpage
+  file(GLOB_RECURSE LANDINGPAGE_SOURCE_FILES RELATIVE ${CMAKE_CURRENT_SOURCE_DIR} server/src/landingpage/*)
+  set(LANDINGPAGE_OUTPUT_PATH ${CMAKE_CURRENT_BINARY_DIR}/../output/data/resources/server/src/landingpage)
```
```suggestion
  set(LANDINGPAGE_OUTPUT_PATH ${CMAKE_BINARY_DIR}/output/data/resources/server/src/landingpage)
```

pathmapper replied:
Thanks!

pathmapper committed:
Use more intuitive name for target

Co-authored-by: Denis Rouzaud <denis.rouzaud@gmail.com>

pathmapper committed:
Update output path

Co-authored-by: Denis Rouzaud <denis.rouzaud@gmail.com>

pathmapper committed:
Delete complied js files

pathmapper committed:
Add yarn to mingw deps

pathmapper commented:
Looks like `yarn` is not available for ci builds. 

**Linux:**
@3nids do we need to add it here?
https://github.com/qgis/QGIS/blob/master/.docker/qgis3-qt5-build-deps.dockerfile

**Windows:**
This didn't work: https://github.com/qgis/QGIS/pull/46643/commits/967a70ce957ca966c209df49a9bcde47c7834b90
Anyone has an idea how to make yarn available?

pathmapper committed:
Add mingw64-yarn

pathmapper committed:
Try to add yarnpkg

m-kuhn commented:
> master/.docker/qgis3-qt5-build-deps.dockerfile

That looks good

pathmapper committed:
Add nodejs and yarn to docker

pathmapper committed:
Migrate to vue cli v5 to solve issue related to webpack 4 and fedora rawhide which already uses openssl 3

m-kuhn commented:
Nice to see that this also helps to improve js compatibility

pathmapper committed:
Update dependencies

pathmapper commented:
The `ubuntu` build and all tests are passing, but the `rawhide` build fails with 

```
  ERROR  Failed to compile with 1 error7:33:27 PM
  
  error  in ./src/main.js
  
  Syntax Error: Thread Loader (Worker 0)
  error:0308010C:digital envelope routines::unsupported
      at Generator.next (<anonymous>)
      at new Promise (<anonymous>)
```

and I'm running out of ideas how to fix this. 

From the error message it looks like an openssl issue but it could also be related to rawhide, how rawhide implements openssl 3, a js dependency or something else...

I've created https://github.com/pathmapper/test-landingpage to test the landingpage ci build on different platforms (the landingpage source is same as used in this PR).

The build works fine on 
- ubuntu-latest (node 16 and 17)
- windows-latest (node 16 and 17)
- macos-latest (node 16 and 17)
- **fedora 35 (node 16)**

and fails for fedora rawhide (node 16) with the same error:

https://github.com/pathmapper/test-landingpage/pull/2

![image](https://user-images.githubusercontent.com/20856381/147811350-1a12eef8-6c7c-4efb-ade1-4a81cfa5c041.png)

I think the options are:

1. Wait until this is fixed in rawhide or somewhere else upstream (e.g. rawhide currently uses OpenSSL 3.0 but [3.1 is released](https://www.openssl.org/news/changelog.html#openssl-31)).
2. Run rawhide mingw CI build without server for the time being.
3. Use fedora 35 for the mingw CI build (this has been raised before, https://github.com/qgis/QGIS/pull/45368).

m-kuhn commented:
Cc @manisandro

manisandro commented:
It's most likely because fedora rawhide uses openssl-3.0.0 which disabled deprecated routines by default. So it's probably not a bug in openssl, but rather something using a deprecated routine (i.e. we hit this in various webpack based node projects, as webpack still uses md4 by default in various places.

Looking at vue.config.js, maybe adding

    module.exports = {
      ...
      configureWebpack: {
        output: {
            hashFunction: 'sha256',
        }
    }

helps?

pathmapper commented:
Thanks @manisandro I did already try this with no success, just double checked here:
https://github.com/pathmapper/test-landingpage/pull/3

> we hit this in various webpack based node projects, as webpack still uses md4 by default in various places

There was a similar but different error before (https://github.com/webpack/webpack/issues/14532) which was fixed here by upgrading to Vue CLI v5  https://github.com/qgis/QGIS/pull/46643/commits/a9ae64e5f2809701d01fc4489a158f1834faec30  which uses a webpack 5 version with some fixes regarding this issue.

Also this reported issue https://github.com/webpack/webpack/issues/14532 happens with node 17 and the node 17 builds in https://github.com/pathmapper/test-landingpage/pull/2 are passing.

pathmapper commented:
Node 17 uses OpenSSL 3 : https://medium.com/the-node-js-collection/node-js-17-is-here-8dba1e14e382

manisandro commented:
So I tried

    find -type f -exec sed -i 's|md4|sha256|g' {} \;

in the node_modules directory and the build succeeds.

pathmapper committed:
Use node 17 for rawhide build

pathmapper committed:
Node.js and yarn are needed for building server landingpage

pathmapper committed:
Use 'corepack enable' to make yarn available

pathmapper committed:
Add comment

pathmapper commented:
Thanks @manisandro, very helpful :tada: 

The culprit was the `babel-loader` dependency which currently still uses md4 hashing for node versions <17. 
https://github.com/babel/babel-loader/blob/052c07a0168a24e2bb15fd8890c9e485c5a3fe97/src/cache.js#L66-L70

So using node 17 for the rawhide build works as expected, all checks are passing now.

Two remarks:

**1.**

With Vue CLI v5, which is used here, by default there are two different bundles created, one which targets modern browsers and one which targets also legacy browsers: 
https://next.cli.vuejs.org/migrations/migrate-from-v4.html#the-build-command-and-modern-mode

**Legacy bundle:**
![legacy_bundle](https://user-images.githubusercontent.com/20856381/147848318-3ef04dd4-4a06-4fb6-93b7-21431ef89d09.png)

**Modern bundle:**
![modern_bundle](https://user-images.githubusercontent.com/20856381/147848324-1b5c65d4-7759-4905-a43b-ffc970d7ee14.png)

Please note the difference in the data size. This is in general nice but roughly doubles the build time, so currently only the legacy bundly is built for the server landingpage. This behaviour could be changed of course, if wanted (build both, or build only the modern bundle).

**2.**

If this PR would be merged, for builds including QGIS server based on this PR **Node.js** and **Yarn** are needed.
For Linux some information was added to the building guide -> https://github.com/qgis/QGIS/pull/46643/commits/fbe0bb017176bf98103fa725da5842cc237ac0c2

For Windows (and MacOS X?) it would be much appreciated if someone with good knowledge of builds on such platforms could add the needed information as a follow-up on this PR.

m-kuhn reviewed (approved):
Thanks for following up @pathmapper and @manisandro 

I think the server is currently mostly thought to be a development kit, so I think it would be ok to only build the modern page. Building the legacy one could be a cmake parameter if wanted (?)

I am happy to merge this as it currently stands and follow up with anything that might come later when it's time. I'll leave it open for a week in case someone has objections or if you want to fine tune things in here @pathmapper

pblottiere commented:
Thanks @pathmapper :). I'm currently cloning your branch to test the build process on Arch Linux :crossed_fingers:.

pathmapper commented:
> I am happy to merge this as it currently stands and follow up with anything that might come later when it's time. I'll leave it open for a week in case someone has objections or if you want to fine tune things in here @pathmapper

Yeah, sounds good. Thanks a lot for reviewing and your input here @m-kuhn .

lbartoletti commented:
> Thanks @pathmapper :). I'm currently cloning your branch to test the build process on Arch Linux crossed_fingers.

OK for FreeBSD.

pblottiere commented:
By trying to compile on a fresh environment, I noticed that the `yarn` binary is not checked when running `cmake`, leading to the next error when compiling:

``` bas...
[ 55%] Built target svg
[ 55%] Generating ../output/data/resources/server/src/landingpage/README.md
[ 55%] Generating ../output/data/resources/server/src/landingpage/babel.config.js
[ 55%] Generating ../output/data/resources/server/src/landingpage/landingpage.stamp
/bin/sh: line 1: yarn: command not found
```

Shouldn't we test if `yarn` is available in `CMakeLists.txt`?

m-kuhn commented:
True, a `find_program` call would be great, that would also give better control than just to trust `PATH`

pblottiere commented:
@pathmapper nice, the build process is operational on Arch Linux and the landing page seems to be working correctly :+1:. 

But I'm wondering if adding a hard constraint on `yarn` for building the server is a good idea considering that it's a web environment. I think I'd prefer to have a dedicated boolean option `WITH_SERVER_LANDING_PAGE` to activate the option. What do you think?

lbartoletti commented:
> But I'm wondering if adding a hard constraint on `yarn` for building the server is a good idea considering that it's a web environment. I think I'd prefer to have a dedicated boolean option `WITH_SERVER_LANDING_PAGE` to activate the option. What do you think?

Same here. Please, add an option to require `yarn`

manisandro commented:
> @pathmapper nice, the build process is operational on Arch Linux and the landing page seems to be working correctly +1.
> 
> But I'm wondering if adding a hard constraint on `yarn` for building the server is a good idea considering that it's a web environment. I think I'd prefer to have a dedicated boolean option `WITH_SERVER_LANDING_PAGE` to activate the option. What do you think?

I think this may also be a good idea for distro packaging. I haven't looked into it in detail yet, but I'd assume that network access is required to build this? If so, this will be a no-go for distro-packaging, since the builders typically disable network access.

pathmapper commented:
Thank you all for testing and feedback. 

Until now there was no dedicated cmake boolean option like `WITH_SERVER_LANDING_PAGE` because my goal for this PR was to integrate the landingpage build with cmake without changing too much and the landingpage was included for all server builds so far.

But such a dedicated option for the server landingpage could be added of course, seems like there's a demand for it :wink: .

> I haven't looked into it in detail yet, but I'd assume that network access is required to build this? 

Yes, with `yarn install` the js dependencies are retrieved via internet.

> If so, this will be a no-go for distro-packaging, since the builders typically disable network access.

Oh, that's sad news. 

@manisandro, you wrote 

> typically

Can someone confirm?  It would mean that we would have a breaking change for the server distro packages because they would come without the server landingpage. Until now the server landingpage is included in distro packages. 

Also there would be a need to build the landingpage yourself if you want to use it, so the entry barrier would be higher.

I think @elpaso has created with the landingpage/catalog an unique feature which adds additional value to QGIS server which other server software currently doesn't have to offer.

I'm not sure where we are standing now. Feels like with this PR landingpage development would become easier along with other advantages (e.g. https://github.com/qgis/QGIS/pull/45802#issuecomment-968806555). On the other hand it would be bad for the adoption of the landingpage if it wouldn't be included in distro packages anymore. 

So maybe close this PR and keep the the situation with the compiled js files in the source tree?

m-kuhn commented:
CC @sebastic @daniviga @jef-n 

Do you know more (or know who knows more) about packaging with yarn dependencies?

Solving packaging problems in this repository by keeping copies of external dependencies sounds like an interesting approach to me. I'm tempted to believe there must be a better place in the stack for this task.

manisandro commented:
You can build a yarn package cache with

    export YARN_CACHE_FOLDER="$(pwd)/.package-cache"
    yarn install --frozen-lockfile

And then delete any precompiled object files. Example shell script (used for Fedora packaging): [prepare_vendor.txt](https://github.com/qgis/QGIS/files/7809264/prepare_vendor.txt)

sebastic commented:
> Do you know more (or know who knows more) about packaging with yarn dependencies?

The Node.js ecosystem is a pain to package, I avoid it like the plague.

If the QGIS build is going to require nodejs, I'm likely to patch it out.

lbartoletti commented:
> The Node.js ecosystem is a pain to package, I avoid it like the plague.
> 
> If the QGIS build is going to require nodejs, I'm likely to patch it out.

+1

>... If so, this will be a no-go for distro-packaging, since the builders typically disable network access.

I confirm

elpaso commented:
@pathmapper how about a github workflow that builds the js files and commits them into the source tree?

manisandro commented:
> > @pathmapper how about a github workflow that builds the js files and commits them into the source tree?
> 
> I was thinking the same but this needs to be in a separate repository that could be used for packaging.

As far as Fedora/Centos/etc is concerned, this will still be problematic as there is a general requirement that all code is compiled from source during the package build process. It would actually be easier to include the yarn cache in the qgis source package and use yarn like in the regular online case rather than having to pick out the precompiled assets and selectively rebuild those. Easiest option would still be the flag to disable the portion of qgis which requires nodejs.

m-kuhn commented:
If this is to stay in the main git, I'd rather have a pre-commit hook / workflow to check if it's updated (like e.g. for sip files) than a workflow that commits into this repo.

pathmapper commented:
I understood @elpaso's suggestion that way, that the workflow builds the js files e.g. on push to the main / release branches (after a PR is merged) and commits them to the source tree. So they would be part of [qgis/QGIS](https://github.com/qgis/QGIS) but there's no need to build and add them for every landingpage PR. Drawback would be that there's a need to take care of failing workflow-runs.

> include the yarn cache in the qgis source package and use yarn like in the regular online case

I understood @sebastic that way, that this would be a no-go for him:

> If the QGIS build is going to require nodejs, I'm likely to patch it out.

manisandro commented:
To clarify, by

> include the yarn cache in the qgis source package and use yarn like in the regular online case

I meant to include the yarn cache on the distro side, not upstream side.

pathmapper commented:
> As far as Fedora/Centos/etc is concerned, this will still be problematic as there is a general requirement that all code is compiled from source during the package build process

But e.g. the current Fedora `qgis-server` packages includes the compiled js files?
https://github.com/qgis/QGIS/tree/master/resources/server/api/ogc/static/landingpage/js

manisandro commented:
That had escaped my attention - per guidelines, I have to remove those files.

pathmapper committed:
Add dedicated cmake boolean option for server landingpage webapp

m-kuhn commented on CMakeLists.txt:
```diff
@@ -197,6 +197,8 @@ if(WITH_CORE)
   # server disabled default because it needs FastCGI (which is optional dependency)
   set (WITH_SERVER FALSE CACHE BOOL "Determines whether QGIS server should be built")
   if(WITH_SERVER)
+    set (WITH_SERVER_LANDINGPAGE_WEBAPP FALSE CACHE BOOL "Determines whether QGIS server landingpage webapp should be built (building webapp disabled by default because it needs nodejs and yarn (which are optional dependencies)")
```
```suggestion
    set (WITH_SERVER_LANDINGPAGE_WEBAPP FALSE CACHE BOOL "Determines whether QGIS server landingpage webapp should be built (requires nodejs and yarn)")
```

pathmapper committed:
Make option description shorter

Co-authored-by: Matthias Kuhn <matthias@opengis.ch>

elpaso commented:
What's the bottom line here? Are we dropping the landing page functionality from packages because of the various policies?

I've put a lot of volunteer effort in that development (nobody paid for it and I am not using it for any of my customers) and I think it would be a pity to drop it: there is actually a nice (possibly underrated) use case where you can store your QGIS projects and data into a remote PG DB and have them web published right away into the landing page catalog using your QGIS desktop with zero efforts, I believe this is a unique selling point for QGIS server.

@manisandro @sebastic do you see any viable option to continue packaging and shipping this functionality?

m-kuhn commented:
The main goal here should be:

- to improve the developer experience when editing the landing page
- especially when upstreaming changes
- to reduce the upstream maintenance effort

This comment https://github.com/qgis/QGIS/pull/46643#issuecomment-1005717454 looks like it's a task that can and should be solved in packaging

sebastic commented:
> @manisandro @sebastic do you see any viable option to continue packaging and shipping this functionality?

Not really. The Node.js bits don't belong in QGIS in my opinion, a desktop GIS shouldn't be building web stacks.

gioman commented:
> I believe this is a unique selling point for QGIS server.

Agree, it would be a pity see this go, both for the functionality (that is really an important selling point) and for the effort that was put into it.

elpaso commented:
> > @manisandro @sebastic do you see any viable option to continue packaging and shipping this functionality?
> 
> Not really. The Node.js bits don't belong in QGIS in my opinion, a desktop GIS shouldn't be building web stacks.

We are talking about QGIS Server packages here, not the desktop.

manisandro commented:
In theory bundling the entire yarn cache with no precompiled/minified code and doing a full license review of all dependencies and listing those in the package license it should be feasible to build the functionality, but that said it's a pretty demanding task and as far as I see in Fedora, package maintainers are giving up on packaging nodejs stuff at distro level.

m-kuhn commented:
@manisandro if I understand you correctly, shipping js code from the QGIS repo (upstream) is not permitted per guidelines and if the landing page should be included in packages it needs to be packaged as you described in https://github.com/qgis/QGIS/pull/46643#issuecomment-1005717454 (but this requires commitment from the packager)?

manisandro commented:
So the relevant portion of the guidelines in the Fedora case are here [1]. The interpretation for JS is that minified code is that it equals to compiled binaries. The issue with JS concerns "just" minified code, not the human readable source.

[1] https://docs.fedoraproject.org/en-US/packaging-guidelines/what-can-be-packaged/#prebuilt-binaries-or-libraries

sebastic commented:
> We are talking about QGIS Server packages here, not the desktop.

The qgis-server package is built from the qgis source, and hence indistinguishable from the desktop parts.

Leave it up to the user to embellish the qgis project with javascript bits, don't include that in qgis itself.

pathmapper commented:
@sebastic did something change since https://github.com/qgis/QGIS/issues/41468#issuecomment-776522044 ?

> If non-minified JS is not an option for the landing page in QGIS, its sources must be included in the QGIS source tree.

Sources are included in the QGIS source tree.

> Ideally building the minified JS files is also part of the QGIS build process.

This is what this PR does.

landryb commented:
thanks @lbartoletti for pointing me to this issue - as for OpenBSD i'm on the same side as @sebastic and @manisandro, network access is disabled when building, and i'd also like to avoid packaging the yarn monstrosity to be able to build/package qgis 3.24 if the yarn cache was to be shipped in the tarball.

cant all those preprocessing/minifying steps be only required for builds from git, and source tarballs provided on https://qgis.org/downloads/ (which, as packagers, is our source of truth) for releases contain the end-user generated bits ? that would work also for me.

elpaso commented:
> @sebastic did something change since [#41468 (comment)](https://github.com/qgis/QGIS/issues/41468#issuecomment-776522044) ?
> 
> > If non-minified JS is not an option for the landing page in QGIS, its sources must be included in the QGIS source tree.
> 
> Sources are included in the QGIS source tree.

I believe that the problem is not the vue sources which I created but all the nodejs dependencies that are used to build the whole JS bundle.

Maybe the solution would be to include a frozen copy of the js dependencies in the QGIS source code and build from that copy instead of pulling them from the network like normally happens (for the first time I mean, subsequent builds use the local cache).

We are probably talking about hundreds of files.

> 
> > Ideally building the minified JS files is also part of the QGIS build process.
> 
> This is what this PR does.

m-kuhn commented:
> cant all those preprocessing/minifying steps be only required for builds from git, and source tarballs provided on [qgis.org/downloads](https://qgis.org/downloads/) (which, as packagers, is our source of truth) for releases contain the end-user generated bits ? that would work also for me.

That sounds pragmatic to me

pathmapper committed:
Test if yarn is available

m-kuhn commented on resources/CMakeLists.txt:
```diff
@@ -36,6 +36,11 @@ endif()
 
 # Server landingpage webapp
 if (WITH_SERVER_LANDINGPAGE_WEBAPP)
+  find_program(YARN yarn)
+  if(NOT YARN)
+    message(FATAL_ERROR "Yarn not found, required for server landingpage webapp!")
+  endif()
+
```
```suggestion
  find_program(YARN yarn REQUIRED)

```

I think that should do

m-kuhn replied:
Now below, use `    COMMAND cd ${LANDINGPAGE_OUTPUT_PATH} && ${YARN} install --frozen-lockfile && ${YARN} build
`

pathmapper replied:
Thanks and done.

pathmapper committed:
Simplify test for yarn

Co-authored-by: Matthias Kuhn <matthias@opengis.ch>

pathmapper committed:
Use variable for command

nyalldawson commented:
> Leave it up to the user to embellish the qgis project with javascript bits, don't include that in qgis itself.

I disagree -- the client component here is very much an integral part of QGIS server! (and is a very important feature in itself)

m-kuhn commented:
What's acceptable / required from a packaging point of view?

1. Minified source code in the git repo and source packages (status quo before this pull request)
2. yarn.lock only, source fetched / compiled as part of make, e.g. yarn cache living in the packaging repo (status quo of this pull request)
3. Yarn cache (unminified) generated on release into source tarballs
4. Yarn cache (unminified) living inside the git repository and source packages
5. Javascript does not belong into packages, do what the heck you want

sebastic commented:
> What's acceptable / required from a packaging point of view?
> 
>     1. Minified source code in the git repo and source packages (status quo before this pull request)

Unminified source is required for packages. The minifying should be done as part of the build (or skipped).

>     2. yarn.lock only, source fetched / compiled as part of make (status quo of this pull request; [yarn cache can still live in the packaging repo](https://github.com/qgis/QGIS/pull/46643#issuecomment-1005717454))

A bunch of Node.js libraries in the QGIS (build) dependency chain makes packagers like me very unhappy.

>     3. Yarn cache (unminified) generated on release into source tarballs
> 
>     4. Yarn cache (unminified) living inside the git repository and source packages

A whole bunch of Node.js library sources in the QGIS source tree makes packages loike me also very unhappy, the license and copyright review will be a big pain.
 
>     5. Javascript does not belong into packages, do what the heck you want

Make the Javascript bits into a plugin that the user can download when he needs it. That leaves it out of the QGIS source tree and (package) builds.

manisandro commented:
On my part:
1:  Not acceptable
2: Yes, this is what is being done for the few nodejs packages which survived
3, 4: Yes, though from a packaging POW not much less work than 2 (still need the full license review of all bundled components = dependencies), so I don't believe it is worth adding the weight of the dependencies to the qgis repo or even tarball.
5. I'd leave the option to disable the nodejs part, depending on the guidelines of the target distro

m-kuhn commented:
Thanks for the responses so far.
Is the license review something that can be shared / reused?

elpaso commented:
@m-kuhn maybe this could be of help: https://www.npmjs.com/package/license-checker

manisandro commented:
> Thanks for the responses so far. Is the license review something that can be shared / reused?

I guess adding a file containing the licenses of all dependencies could be useful yes, if there is some automated way of generating it and ensuring it stays up-to-date.

pathmapper commented:
According to it's description, https://github.com/codepunkt/webpack-license-plugin should  

> - Discover every npm package used in your webpack output
> - Find out how it is licensed

Just tried it with the current status of this PR and when building, along with the minified js files this plugin creates also a json file 

![image](https://user-images.githubusercontent.com/20856381/148600635-68afb771-7a0f-4d67-a4cc-f5b61804cb67.png)


with the following content:
 
<details>
  <summary>show "oss-licenses.json"</summary>
  
```json
[
  {
    "name": "@babel/runtime",
    "version": "7.16.5",
    "author": "The Babel Team (https://babel.dev/team)",
    "repository": "https://github.com/babel/babel",
    "source": "https://registry.npmjs.org/@babel/runtime/-/runtime-7.16.5.tgz",
    "license": "MIT",
    "licenseText": "MIT License\n\nCopyright (c) 2014-present Sebastian McKenzie and other contributors\n\nPermission is hereby granted, free of charge, to any person obtaining\na copy of this software and associated documentation files (the\n\"Software\"), to deal in the Software without restriction, including\nwithout limitation the rights to use, copy, modify, merge, publish,\ndistribute, sublicense, and/or sell copies of the Software, and to\npermit persons to whom the Software is furnished to do so, subject to\nthe following conditions:\n\nThe above copyright notice and this permission notice shall be\nincluded in all copies or substantial portions of the Software.\n\nTHE SOFTWARE IS PROVIDED \"AS IS\", WITHOUT WARRANTY OF ANY KIND,\nEXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF\nMERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND\nNONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE\nLIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION\nOF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION\nWITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.\n"
  },
  {
    "name": "core-js",
    "version": "3.20.1",
    "repository": "https://github.com/zloirock/core-js",
    "source": "https://registry.npmjs.org/core-js/-/core-js-3.20.1.tgz",
    "license": "MIT",
    "licenseText": "Copyright (c) 2014-2021 Denis Pushkarev\n\nPermission is hereby granted, free of charge, to any person obtaining a copy\nof this software and associated documentation files (the \"Software\"), to deal\nin the Software without restriction, including without limitation the rights\nto use, copy, modify, merge, publish, distribute, sublicense, and/or sell\ncopies of the Software, and to permit persons to whom the Software is\nfurnished to do so, subject to the following conditions:\n\nThe above copyright notice and this permission notice shall be included in\nall copies or substantial portions of the Software.\n\nTHE SOFTWARE IS PROVIDED \"AS IS\", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR\nIMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,\nFITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE\nAUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER\nLIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,\nOUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN\nTHE SOFTWARE.\n"
  },
  {
    "name": "leaflet",
    "version": "1.7.1",
    "repository": "https://github.com/Leaflet/Leaflet",
    "source": "https://registry.npmjs.org/leaflet/-/leaflet-1.7.1.tgz",
    "license": "BSD-2-Clause",
    "licenseText": "Copyright (c) 2010-2019, Vladimir Agafonkin\nCopyright (c) 2010-2011, CloudMade\nAll rights reserved.\n\nRedistribution and use in source and binary forms, with or without modification, are\npermitted provided that the following conditions are met:\n\n   1. Redistributions of source code must retain the above copyright notice, this list of\n      conditions and the following disclaimer.\n\n   2. Redistributions in binary form must reproduce the above copyright notice, this list\n      of conditions and the following disclaimer in the documentation and/or other materials\n      provided with the distribution.\n\nTHIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS \"AS IS\" AND ANY\nEXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF\nMERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE\nCOPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,\nEXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF\nSUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)\nHOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR\nTORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS\nSOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.\n"
  },
  {
    "name": "leaflet-wms",
    "version": "0.1.0",
    "repository": null,
    "source": "https://registry.npmjs.org/leaflet-wms/-/leaflet-wms-0.1.0.tgz",
    "license": "MIT",
    "licenseText": "The MIT License (MIT)\n\nCopyright (c) 2014 Houston Engineering, Inc.\n\nPermission is hereby granted, free of charge, to any person obtaining a copy of\nthis software and associated documentation files (the \"Software\"), to deal in\nthe Software without restriction, including without limitation the rights to\nuse, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of\nthe Software, and to permit persons to whom the Software is furnished to do so,\nsubject to the following conditions:\n\nThe above copyright notice and this permission notice shall be included in all\ncopies or substantial portions of the Software.\n\nTHE SOFTWARE IS PROVIDED \"AS IS\", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR\nIMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS\nFOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR\nCOPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER\nIN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN\nCONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.\n"
  },
  {
    "name": "proj4",
    "version": "2.7.5",
    "author": "",
    "repository": "https://github.com/proj4js/proj4js",
    "source": "https://registry.npmjs.org/proj4/-/proj4-2.7.5.tgz",
    "license": "MIT",
    "licenseText": "## Proj4js -- Javascript reprojection library. \n \nAuthors:\n- Mike Adair madairATdmsolutions.ca\n- Richard Greenwood richATgreenwoodmap.com\n- Didier Richard didier.richardATign.fr\n- Stephen Irons stephen.ironsATclear.net.nz\n- Olivier Terral oterralATgmail.com\n- Calvin Metcalf cmetcalfATappgeo.com\n\nCopyright (c) 2014, Mike Adair, Richard Greenwood, Didier Richard, Stephen Irons, Olivier Terral and Calvin Metcalf\n\n Permission is hereby granted, free of charge, to any person obtaining a\n copy of this software and associated documentation files (the \"Software\"),\n to deal in the Software without restriction, including without limitation\n the rights to use, copy, modify, merge, publish, distribute, sublicense,\n and/or sell copies of the Software, and to permit persons to whom the\n Software is furnished to do so, subject to the following conditions:\n\n The above copyright notice and this permission notice shall be included\n in all copies or substantial portions of the Software.\n\n _THE SOFTWARE IS PROVIDED \"AS IS\", WITHOUT WARRANTY OF ANY KIND, EXPRESS\n OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,\n FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL\n THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER\n LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING\n FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER\n DEALINGS IN THE SOFTWARE._\n"
  },
  {
    "name": "regenerator-runtime",
    "version": "0.13.9",
    "author": "Ben Newman <bn@cs.stanford.edu>",
    "repository": "https://github.com/facebook/regenerator/tree/master/packages/runtime",
    "source": "https://registry.npmjs.org/regenerator-runtime/-/regenerator-runtime-0.13.9.tgz",
    "license": "MIT",
    "licenseText": "MIT License\n\nCopyright (c) 2014-present, Facebook, Inc.\n\nPermission is hereby granted, free of charge, to any person obtaining a copy\nof this software and associated documentation files (the \"Software\"), to deal\nin the Software without restriction, including without limitation the rights\nto use, copy, modify, merge, publish, distribute, sublicense, and/or sell\ncopies of the Software, and to permit persons to whom the Software is\nfurnished to do so, subject to the following conditions:\n\nThe above copyright notice and this permission notice shall be included in all\ncopies or substantial portions of the Software.\n\nTHE SOFTWARE IS PROVIDED \"AS IS\", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR\nIMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,\nFITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE\nAUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER\nLIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,\nOUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE\nSOFTWARE.\n"
  },
  {
    "name": "reproject",
    "version": "1.2.6",
    "author": "Per Liedman",
    "repository": null,
    "source": "https://registry.npmjs.org/reproject/-/reproject-1.2.6.tgz",
    "license": "MIT",
    "licenseText": "Copyright (c) 2013 Per Liedman\nAll rights reserved.\n\nRedistribution and use in source and binary forms, with or without modification, are\npermitted provided that the following conditions are met:\n\n   1. Redistributions of source code must retain the above copyright notice, this list of\n      conditions and the following disclaimer.\n\n   2. Redistributions in binary form must reproduce the above copyright notice, this list\n      of conditions and the following disclaimer in the documentation and/or other materials\n      provided with the distribution.\n\nTHIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS \"AS IS\" AND ANY\nEXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF\nMERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE\nCOPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,\nEXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF\nSUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)\nHOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR\nTORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS\nSOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE."
  },
  {
    "name": "uuid",
    "version": "8.3.2",
    "repository": "https://github.com/uuidjs/uuid",
    "source": "https://registry.npmjs.org/uuid/-/uuid-8.3.2.tgz",
    "license": "MIT",
    "licenseText": "The MIT License (MIT)\n\nCopyright (c) 2010-2020 Robert Kieffer and other contributors\n\nPermission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the \"Software\"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:\n\nThe above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.\n\nTHE SOFTWARE IS PROVIDED \"AS IS\", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.\n"
  },
  {
    "name": "vue",
    "version": "2.6.14",
    "author": "Evan You",
    "repository": "https://github.com/vuejs/vue",
    "source": "https://registry.npmjs.org/vue/-/vue-2.6.14.tgz",
    "license": "MIT",
    "licenseText": "The MIT License (MIT)\n\nCopyright (c) 2013-present, Yuxi (Evan) You\n\nPermission is hereby granted, free of charge, to any person obtaining a copy\nof this software and associated documentation files (the \"Software\"), to deal\nin the Software without restriction, including without limitation the rights\nto use, copy, modify, merge, publish, distribute, sublicense, and/or sell\ncopies of the Software, and to permit persons to whom the Software is\nfurnished to do so, subject to the following conditions:\n\nThe above copyright notice and this permission notice shall be included in\nall copies or substantial portions of the Software.\n\nTHE SOFTWARE IS PROVIDED \"AS IS\", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR\nIMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,\nFITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE\nAUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER\nLIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,\nOUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN\nTHE SOFTWARE.\n"
  },
  {
    "name": "vue-loader",
    "version": "15.9.8",
    "author": "Evan You",
    "repository": "https://github.com/vuejs/vue-loader",
    "source": "https://registry.npmjs.org/vue-loader/-/vue-loader-15.9.8.tgz",
    "license": "MIT",
    "licenseText": "The MIT License (MIT)\n\nCopyright (c) 2015-present Yuxi (Evan) You\n\nPermission is hereby granted, free of charge, to any person obtaining a copy\nof this software and associated documentation files (the \"Software\"), to deal\nin the Software without restriction, including without limitation the rights\nto use, copy, modify, merge, publish, distribute, sublicense, and/or sell\ncopies of the Software, and to permit persons to whom the Software is\nfurnished to do so, subject to the following conditions:\n\nThe above copyright notice and this permission notice shall be included in all\ncopies or substantial portions of the Software.\n\nTHE SOFTWARE IS PROVIDED \"AS IS\", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR\nIMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,\nFITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE\nAUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER\nLIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,\nOUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE\nSOFTWARE.\n"
  },
  {
    "name": "vue-router",
    "version": "3.5.3",
    "author": "Evan You",
    "repository": "https://github.com/vuejs/vue-router",
    "source": "https://registry.npmjs.org/vue-router/-/vue-router-3.5.3.tgz",
    "license": "MIT",
    "licenseText": "MIT License\n\nCopyright (c) 2013-present Evan You\n\nPermission is hereby granted, free of charge, to any person obtaining a copy\nof this software and associated documentation files (the \"Software\"), to deal\nin the Software without restriction, including without limitation the rights\nto use, copy, modify, merge, publish, distribute, sublicense, and/or sell\ncopies of the Software, and to permit persons to whom the Software is\nfurnished to do so, subject to the following conditions:\n\nThe above copyright notice and this permission notice shall be included in all\ncopies or substantial portions of the Software.\n\nTHE SOFTWARE IS PROVIDED \"AS IS\", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR\nIMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,\nFITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE\nAUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER\nLIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,\nOUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE\nSOFTWARE.\n"
  },
  {
    "name": "vue2-leaflet",
    "version": "2.7.1",
    "author": "Mickael Bouchaud",
    "repository": "https://github.com/vue-leaflet/Vue2Leaflet",
    "source": "https://registry.npmjs.org/vue2-leaflet/-/vue2-leaflet-2.7.1.tgz",
    "license": "MIT",
    "licenseText": "The MIT License (MIT)\n\nCopyright (c) 2016-2017 Mickael Bouchaud\n\nPermission is hereby granted, free of charge, to any person obtaining a copy\nof this software and associated documentation files (the \"Software\"), to deal\nin the Software without restriction, including without limitation the rights\nto use, copy, modify, merge, publish, distribute, sublicense, and/or sell\ncopies of the Software, and to permit persons to whom the Software is\nfurnished to do so, subject to the following conditions:\n\nThe above copyright notice and this permission notice shall be included in\nall copies or substantial portions of the Software.\n\nTHE SOFTWARE IS PROVIDED \"AS IS\", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR\nIMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,\nFITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE\nAUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER\nLIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,\nOUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN\nTHE SOFTWARE.\n"
  },
  {
    "name": "vuetify",
    "version": "2.6.2",
    "author": "John Leider <john@vuetifyjs.com>",
    "repository": "https://github.com/vuetifyjs/vuetify",
    "source": "https://registry.npmjs.org/vuetify/-/vuetify-2.6.2.tgz",
    "license": "MIT",
    "licenseText": "The MIT License (MIT)\n\nCopyright (c) 2016-2020 John Jeremy Leider\n\nPermission is hereby granted, free of charge, to any person obtaining a copy\nof this software and associated documentation files (the \"Software\"), to deal\nin the Software without restriction, including without limitation the rights\nto use, copy, modify, merge, publish, distribute, sublicense, and/or sell\ncopies of the Software, and to permit persons to whom the Software is\nfurnished to do so, subject to the following conditions:\n\nThe above copyright notice and this permission notice shall be included in\nall copies or substantial portions of the Software.\n\nTHE SOFTWARE IS PROVIDED \"AS IS\", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR\nIMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,\nFITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE\nAUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER\nLIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,\nOUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN\nTHE SOFTWARE.\n"
  },
  {
    "name": "vuetify-loader",
    "version": "1.7.3",
    "author": "",
    "repository": "https://github.com/vuetifyjs/vuetify-loader",
    "source": "https://registry.npmjs.org/vuetify-loader/-/vuetify-loader-1.7.3.tgz",
    "license": "MIT",
    "licenseText": "The MIT License (MIT)\n\nCopyright (c) 2016-2019 John Jeremy Leider\n\nPermission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the \"Software\"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:\n\nThe above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.\n\nTHE SOFTWARE IS PROVIDED \"AS IS\", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.\n"
  },
  {
    "name": "vuex",
    "version": "3.6.2",
    "author": "Evan You",
    "repository": "https://github.com/vuejs/vuex",
    "source": "https://registry.npmjs.org/vuex/-/vuex-3.6.2.tgz",
    "license": "MIT",
    "licenseText": "The MIT License (MIT)\n\nCopyright (c) 2015-present Evan You\n\nPermission is hereby granted, free of charge, to any person obtaining a copy\nof this software and associated documentation files (the \"Software\"), to deal\nin the Software without restriction, including without limitation the rights\nto use, copy, modify, merge, publish, distribute, sublicense, and/or sell\ncopies of the Software, and to permit persons to whom the Software is\nfurnished to do so, subject to the following conditions:\n\nThe above copyright notice and this permission notice shall be included in\nall copies or substantial portions of the Software.\n\nTHE SOFTWARE IS PROVIDED \"AS IS\", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR\nIMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,\nFITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE\nAUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER\nLIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,\nOUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN\nTHE SOFTWARE.\n"
  }
]
```

</details>

According to the plugin output there are in total currently 15 dependencies for the minified code with the following licenses:

- 14 x MIT
- 1 x BSD-2-Clause

Does this help? Maybe for packagers it would be enough to add the unminified code / yarn cache of these dependencies listed in the plugin output file.

If yes, we could add the plugin to our config (might make sense anyway).

manisandro commented:
The oss-licenses.json is useful yes, I'd personally leave it at that (=just including that file) as far as the qgis repo is concerned, I feel that the preferred handling of the dependencies might be distro / packager specific.

pathmapper committed:
Create bundle license file when building

pathmapper committed:
Drop support for IE11 and build only modern bundle

IE11 will be retired soon: https://blogs.windows.com/windowsexperience/2021/05/19/the-future-of-internet-explorer-on-windows-10-is-in-microsoft-edge/

pathmapper committed:
Update leaflet.wms dependency

m-kuhn commented:
@elpaso is this mergable?

elpaso commented:
@pathmapper nice work!

elpaso commented:
@pathmapper I'm testing locally `-DWITH_SERVER_LANDINGPAGE_WEBAPP:BOOL=ON` and cmake hangs forever.

Yarn and nodejs are installed.

pathmapper commented:
@elpaso `cmake -DWITH_SERVER:BOOL=ON -DWITH_SERVER_LANDINGPAGE_WEBAPP:BOOL=ON ..` is working fine for me locally on Ubuntu 20.04.3 LTS and the build succeeded.

Do you have a `node_modules` directory in your qgis source tree (`QGIS/resources/server/src/landingpage`) ?
If yes, I would assume the hang would be resolved if you would delete the `node_modules` directory.

elpaso commented:
@pathmapper thank you, that fixed the issue.

pathmapper commented:
@elpaso cool, should I submit a PR to remove these entries from `.gitignore` so it's noticeable with e.g. `git status` / `git diff` that these werde added to the source tree (probably by executing `yarn install` / `build` locally)?

https://github.com/qgis/QGIS/blob/24a498b57cc39c7ef2ddcb8bf11eaab7c9d05a8e/.gitignore#L59-L60

elpaso commented:
> @elpaso cool, should I submit a PR to remove these entries from `.gitignore` so it's noticeable with e.g. `git status` / `git diff` that these werde added to the source tree (probably by executing `yarn install` / `build` locally)?
> 
> https://github.com/qgis/QGIS/blob/24a498b57cc39c7ef2ddcb8bf11eaab7c9d05a8e/.gitignore#L59-L60

Yes, maybe a good idea.
